clients/da-vinci-client/src/main/java/com/linkedin/davinci/store/rocksdb/RocksDBStoragePartition.java

```diff
@@ -548,6 +552,79 @@ public byte[] get(ByteBuffer keyBuffer) {
     }
   }
 
+  public List<byte[]> multiGet(List<byte[]> keys) {
+    readCloseRWLock.readLock().lock();
+    try {
+      makeSureRocksDBIsStillOpen();
+      return rocksDB.multiGetAsList(keys);
+    } catch (RocksDBException e) {
+      throw new VeniceException("Failed to get value from store: " + storeName + ", partition id: " + partitionId, e);
+    } finally {
+      readCloseRWLock.readLock().unlock();
+    }
+  }
+
+  public List<ByteBuffer> multiGet(List<ByteBuffer> keys, List<ByteBuffer> values) {
+    readCloseRWLock.readLock().lock();
+    try {
+      makeSureRocksDBIsStillOpen();
+      List<ByteBufferGetStatus> statusList = rocksDB.multiGetByteBuffers(keys, values);
+
+      List<ByteBuffer> resultList = new ArrayList<>(keys.size());
+      Iterator<ByteBufferGetStatus> statusIter = statusList.iterator();
+      Iterator<ByteBuffer> keyIter = keys.iterator();
+      ListIterator<ByteBuffer> valueIter = values.listIterator();
+      while (keyIter.hasNext()) {
+        if (!statusIter.hasNext() || !valueIter.hasNext()) {
+          throw new VeniceException(
+              "RocksDB returns inconsistent number of result, key count: " + keys.size() + ", but returns: "
+                  + Math.min(statusList.size(), values.size()) + " in store: " + storeName + ", partition: "
+                  + partitionId);
+        }
+        ByteBuffer key = keyIter.next();
+        ByteBufferGetStatus bbStatus = statusIter.next();
+        Status.Code statusCode = bbStatus.status.getCode();
+        ByteBuffer value = valueIter.next();
+        if (statusCode.equals(Status.Code.Ok)) {
+          // Need to check whether the result is complete or not by comparing length
+          if (value.remaining() == bbStatus.requiredSize) {
+            // good
+            resultList.add(value);
+          } else {
+            // Need to look it up again since the passed buffer is too small
+            byte[] newValue;
+            if (key.isDirect()) {
+              byte[] keyBytes = new byte[key.remaining()];
+              key.mark();
+              key.get(keyBytes, 0, key.remaining());
+              key.reset();
+              newValue = get(keyBytes);
```

Comment:
My understanding that the buffer is allocated only because there is no `RocksDB::get` overload that takes `ByteBuffer`? It can be expensive especially if this is not a one off large value. Depending on the number of failed keys it might be faster to retry `multiGetByteBuffers` for them. What do you think?

Reply:
Also since this fallback is expensive, it might be worth having a metric to detect if we're losing performance here.

Reply:
Fallback is expensive and in practice, it should rarely happen since by default, I think we will allocate 1MB buffer and metrics can be expensive in RocksDB layer.
That is the same behavior in singel-get API.